### PR TITLE
doc: add unittest core dump instructions

### PR DESCRIPTION
**Pull request by jamesob**
Quick note on how to get core dumps out of the unittests.

### CODE REVIEW

File: src/test/README.md [hunk 1]
```diff
@@ -74,3 +74,20 @@ start debugging, just like you would with any other program:
...
 ```
+
+#### Segmentation faults
+
+By default, the boost test runner will intercept system errors and not produce a core
+file. To bypass this, add `--catch_system_errors=no` to the `test_bitcoin` arguments
+and ensure that your ulimits are set properly (e.g. `ulimit -c unlimited`). 
+
+Running the tests and hitting a segmentation fault should now produce a file called `core`
+(depending upon your `/proc/sys/kernel/core_pattern` contents).
```
Comment (practicalswift) [inline src/test/README.md:85]: Non-blocking nit: Could clarify that ``/proc/sys/kernel/core_pattern` is Linux specific :)

**Review by practicalswift — COMMENTED** [review]
Concept ACK

File: src/test/README.md (hunk 1 above)
Comment (jamesob) [reply src/test/README.md]: Fixed, thanks

**Review by theStack — COMMENTED** [review]
Concept ACK

Possible follow-ups:
* move 2nd and 3rd paragraph into more general documentation section; IMHO knowing how to find causes for crashes is useful not only for the unit-tests, but also for all other binaries that potentially segfault
* ~~also describe how to analyze core-dumps via `lldb` (if clang is used rather than gcc)~~

**Comment by jamesob** [comment]
> describe how to analyze core-dumps via lldb (if clang is used rather than gcc)

FWIW gdb analyzes clang-built core dumps just fine for me, but maybe you meant something different.

**Review by theStack — APPROVED** [review]
> > describe how to analyze core-dumps via lldb (if clang is used rather than gcc)
> 
> FWIW gdb analyzes clang-built core dumps just fine for me, but maybe you meant something different.

Oh, it seems I had an ancient gdb version installed on my system, that didn't support the right DWARF version (leading me to the wrong conclusion that for clang-built core dump analysis only lldb works -- d'oh!). 

ACK bbeec50921c2cfcf054ba843bda392d310de5782 📜

**Comment by jamesob** [comment]
Okay for merge?

File: src/test/README.md [hunk 2]
```diff
@@ -74,3 +74,19 @@ start debugging, just like you would with any other program:
 ```bash
 gdb src/test/test_bitcoin
 ```
+
+#### Segmentation faults
+
+By default, the boost test runner will intercept system errors and not produce a core
```
Comment (jonatack) [inline src/test/README.md:80]:
```suggestion
By default, the Boost test runner will intercept system errors and not produce a core
```

**Review by jonatack — COMMENTED** [review]
Thanks for adding this. Testing the instructions.

File: src/test/README.md [hunk 3]
```diff
@@ -74,3 +74,19 @@ start debugging, just like you would with any other program:
...
+file. To bypass this, add `--catch_system_errors=no` to the `test_bitcoin` arguments
+and ensure that your ulimits are set properly (e.g. `ulimit -c unlimited`).
+
+Running the tests and hitting a segmentation fault should now produce a file called `core`
+depending upon your `/proc/sys/kernel/core_pattern` contents (on Linux platforms).
```
Comment (jonatack) [inline src/test/README.md:85]:
```suggestion
(on Linux platforms, the file name will depend on the contents of `/proc/sys/kernel/core_pattern`).
```

**Review by jonatack — COMMENTED** [review]
ACK with two suggestions (one above, one below)

**Comment by MarcoFalke** [comment]
review ACK bbeec50921c2cfcf054ba843bda392d310de5782

Personally I just use valgrind which will print the stack when the dump is created

**Comment by jamesob** [comment]
Ready for merge? Or I can add a note about valgrind?

File: src/test/README.md [hunk 4]
```diff
@@ -74,3 +74,19 @@ start debugging, just like you would with any other program:
 ```bash
 gdb src/test/test_bitcoin
```
Comment (MarcoFalke) [inline src/test/README.md:75]: On a second thought. Won't this be enough to observe segmentation faults in gdb if they appear? Is there a reason to take the extra step of a core dump?

Comment (jamesob) [reply src/test/README.md:75]: Yeah, this is a good point; willfix.

**Comment by jamesob** [comment]
Incorporated feedback from @MarcoFalke and @jonatack.

**Comment by practicalswift** [comment]
cr ACK 3ad9b71f48d2783b534264f53e059db4f239ea55

File: src/test/README.md [hunk 5]
```diff
@@ -74,3 +74,29 @@ start debugging, just like you would with any other program:
...
+
+If you hit a segmentation fault during a test run, you can diagnose where the fault
+is happening by running `gdb ./src/test/test_bitcoin` and then using the `bt` command
+within gdb.
+
+Another tool that can be used to resolve segmentation faults is
+[valgrind](https://valgrind.org/).
+
+If for whatever reason you want to produce a core dump file for this fault, you can do
+that as well. By default, the boost test runner will intercept system errors and not
```
Comment (jonatack) [inline src/test/README.md:88]:
```suggestion
that as well. By default, the Boost test runner will intercept system errors and not
```

File: src/test/README.md [hunk 6]
```diff
@@ -74,3 +74,29 @@ start debugging, just like you would with any other program:
...
+produce a core file. To bypass this, add `--catch_system_errors=no` to the
+`test_bitcoin` arguments and ensure that your ulimits are set properly (e.g. `ulimit -c
+unlimited`).
+
+Running the tests and hitting a segmentation fault should now produce a file called `core`
+(on Linux platforms, the file name will likely depend on the contents of
```
Comment (jonatack) [inline src/test/README.md:94]: linter appeasement
```suggestion
(on Linux platforms, the file name will likely depend on the contents of
```

File: src/test/README.md [hunk 7]
```diff
@@ -74,3 +74,29 @@ start debugging, just like you would with any other program:
...
+`test_bitcoin` arguments and ensure that your ulimits are set properly (e.g. `ulimit -c
+unlimited`).
+
+Running the tests and hitting a segmentation fault should now produce a file called `core`
+(on Linux platforms, the file name will likely depend on the contents of 
+`/proc/sys/kernel/core_pattern`).
+
+You can then explore the core dump using
+``` bash
+gdb src/test/test_bitcoin core
```
Comment (jonatack) [inline src/test/README.md:99]:
```suggestion
gdb ./src/test/test_bitcoin core
```

**Review by jonatack — COMMENTED** [review]
This seems like a helpful improvement.

re-ACK modulo suggestions

File: src/test/README.md (hunk 7 above)
Comment (jonatack) [reply src/test/README.md:99]: (you write `gdb ./src/test/test_bitcoin` above. note that this is the same instruction as at the top of the diff)

**Review by theStack — COMMENTED** [review]
almost re-ACK (linter is not happy yet)

Also I agree that the other suggestions by jonatack above (https://github.com/bitcoin/bitcoin/pull/22226#discussion_r678116566, https://github.com/bitcoin/bitcoin/pull/22226#discussion_r678119277) should be tackled.

**Comment by laanwj** [comment]
LGTM, but yes it's failing the linters.

**Comment by jamesob** [comment]
Fixed, thanks for the looks.

**Review by theStack — APPROVED** [review]
ACK 12313382e60c84f106127566d004c03384ca5abf

**Comment by fanquake** [comment]
This has been merged, but wasn't closed by GitHub.

**Pull request closed without merging**